### PR TITLE
Fix iostreams with imbued locales to print INFs correctly

### PR DESCRIPTION
**Pull request by fsb4000**
Followup to #3868.

### CODE REVIEW

**Comment by fsb4000** [comment]
@StephanTLavavej

>Fortunately the fix should be very easy, thanks for checking
>
﻿
>Unhandled exception at 0x00007FFA41F201DD (msvcp140_oss.dll) in TestCpp.exe: Stack cookie instrumentation code detected a stack-based buffer overrun. 
>
﻿
﻿:scream_cat:﻿ 
﻿
﻿It seems, the bug is not related to changes.
﻿
﻿I can repro the error in VS 17.6.3 without my changes at all and with the code:
﻿
```
﻿#include <sstream>
#include <locale>

struct space_out : std::numpunct<char>
{
    char do_thousands_sep()   const { return ' '; }  // separate with spaces
    std::string do_grouping() const { return "\1"; } // groups of 1 digit
};

int main()
{
    std::ostringstream s;
    s.imbue(std::locale(s.getloc(), new space_out));
    s <<12345678;
}

```
﻿
```
﻿cl "/Od" "/EHsc"  /D_ITERATOR_DEBUG_LEVEL=1 /MD main.cpp
```

﻿if we change `/D_ITERATOR_DEBUG_LEVEL=1` to `/D_ITERATOR_DEBUG_LEVEL=0` or `/MD` to `/MDd`, it works.
﻿
﻿
изображение

изображение

But it looks like a library bug because clang-cl also gives the error

```
clang-cl "/Od" "/EHsc"  /D_ITERATOR_DEBUG_LEVEL=1 /MD main.cpp
main
echo %errorlevel%
-1073740791
```
﻿

**Comment by strega-nil-ms** [comment]
Ooh, that's very scary...

File: stl/inc/xlocnum [hunk 1]
```diff
@@ -1376,7 +1376,7 @@ protected:
         const auto _Ngen = static_cast<size_t>(_CSTD sprintf_s(
             &_Buf[0], _Buf.size(), _Ffmt(_Fmt, 0, _Iosbase.flags()), static_cast<int>(_Precision), _Val));
 
-        return _Fput_v2(_Dest, _Iosbase, _Fill, _Buf.c_str(), _Ngen, (_STD isnan)(_Val));
+        return _Fput_v2(_Dest, _Iosbase, _Fill, _Buf.c_str(), _Ngen, (_STD isnan)(_Val) || (_STD isinf)(_Val));
```
Comment (frederick-vs-ja) [inline stl/inc/xlocnum:1379]: How about [`std::isfinite`](https://en.cppreference.com/w/cpp/numeric/math/isfinite)?

**Comment by StephanTLavavej** [comment]
With a custom facet, you can't customize IDL. There's a matrix for this.

File: tests/std/tests/GH_003867_output_nan/test.cpp [hunk 2]
```diff
@@ -5,16 +5,32 @@
 #include <limits>
 #include <locale>
 #include <sstream>
+#include <string>
 
 using namespace std;
 
+struct groups_of_1 : numpunct<char> {
+    // The standard char specialization of std::numpunct::do_thousands_sep returns ','
+    string do_grouping() const {
```
Comment (StephanTLavavej) [inline tests/std/tests/GH_003867_output_nan/test.cpp:14]: This is `virtual`, so we should mark it as `override`.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I've pushed changes to:

* Use `isfinite` as suggested by @frederick-vs-ja, which is faster and simpler than testing `isnan` and `isinf` separately,
* Rename `_Fput_v2` to `_Fput_v3` because the sense of the `bool` is flipping (this isn't really necessary if `_Fput_v2` never shipped for production usage, but it was the cleanest thing to do),
* Guard the test with `TEST_CUSTOM_FACET`, with a comment explaining why,
* Update `Dev11_0496153_locale_ctor` which had an equivalent guard that was phrased in a much more confusing manner. (Its comment previously explained the preprocessor logic in English, but not *why* it existed.)

**Comment by StephanTLavavej** [comment]
D'oh, I checked `std` but didn't rerun the libcxx failures.

**Review by StephanTLavavej — APPROVED** [review]

**Review by strega-nil-ms — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
:infinity: :globe_with_meridians: :tada:
===

**Pull request merged**